contracts/VeJoeStaking.sol

```diff
@@ -0,0 +1,267 @@
+
+            // User is eligible for boosted benefits if:
+            // 1. They are not already currently receiving boosted benefits
+            // 2. Their staked JOE is at least `boostedThreshold / 100 * totalStakedJoe`
+            if (userInfos[msg.sender].boostEndTimestamp == 0) {
+                uint256 totalStakedJoe = joe.balanceOf(address(this));
+                if ((userInfos[msg.sender].balance * 100) / boostedThreshold >= totalStakedJoe) {
+                    userInfos[msg.sender].boostEndTimestamp = block.timestamp + boostedDuration;
+                }
+            }
+        } else {
+            // If the user's `lastRewardTimestamp` is 0, i.e. if this is the user's first time staking,
+            // then they will receive boosted benefits.
+            // Note that it is important we perform this check **before** we update the user's `lastRewardTimestamp`
+            // down below.
+            if (userInfos[msg.sender].lastRewardTimestamp == 0) {
+                userInfos[msg.sender].boostEndTimestamp = block.timestamp + boostedDuration;
+            }
+            userInfos[msg.sender].balance = _amount;
+            userInfos[msg.sender].lastRewardTimestamp = block.timestamp;
+        }
+
+        joe.safeTransferFrom(msg.sender, address(this), _amount);
+
+        emit Deposit(msg.sender, _amount);
+    }
+
+    /// @notice Withdraw staked JOE. Note that unstaking any amount of JOE means you will
+    /// lose all of your current veJOE.
+    /// @param _amount the amount of JOE to unstake
+    function withdraw(uint256 _amount) external {
+        require(_amount > 0, "VeJoeStaking: expected to withdraw non-zero amount of JOE");
+
+        UserInfo storage userInfo = userInfos[msg.sender];
+
+        require(
+            userInfo.balance >= _amount,
+            "VeJoeStaking: cannot withdraw greater amount of JOE than currently staked"
+        );
+
+        userInfo.balance -= _amount;
+        userInfo.lastRewardTimestamp = block.timestamp;
+
+        // Burn the user's current veJOE balance
+        uint256 userVeJoeBalance = veJoe.balanceOf(msg.sender);
+        veJoe.burnFrom(msg.sender, userVeJoeBalance);
+
+        // Send user their requested amount of staked JOE
+        joe.safeTransfer(msg.sender, _amount);
+
+        emit Withdraw(msg.sender, _amount);
+    }
+
+    /// @notice Claim any pending veJOE
+    function claim() external {
+        require(getUserHasNonZeroBalance(msg.sender), "VeJoeStaking: cannot claim any veJOE when no JOE is staked");
+        _claim();
+    }
+
+    /// @notice Get the pending amount of veJOE for a given user
+    /// @param _user The user to lookup
+    /// @return The number of pending veJOE tokens for `_user`
+    function getPendingVeJoe(address _user) public view returns (uint256) {
+        if (!getUserHasNonZeroBalance(_user)) {
+            return 0;
+        }
+
+        UserInfo storage user = userInfos[_user];
```

Comment:
You should use memory here as this is a view.

Reply:
You could also just reference directly like in some other functions and not use the `UserInfo user` variable.